### PR TITLE
[FIRRTL] Add ObjectSubfieldOp

### PR DESCRIPTION
**Pull request by rwy7**
This op indexes into objects (values of type ClassType).

### CODE REVIEW

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp [hunk 1]
```diff
@@ -3932,6 +3932,135 @@ FIRRTLType MultibitMuxOp::inferReturnType(ValueRange operands,
...
+      parser.parseType(inputType))
+    return failure();
+
+  if (parser.resolveOperand(input, inputType, result.operands))
+    return failure();
+
+  auto classType = dyn_cast<ClassType>(inputType);
+  if (!classType)
+    return parser.emitError(parser.getNameLoc(),
+                            "input must be !firrtl.instance type, got ")
```
Comment (dtzSiFive) [inline lib/Dialect/FIRRTL/FIRRTLOps.cpp:4011]:
```suggestion
                            "input must be !firrtl.class type, got ")
```

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp [hunk 2]
```diff
@@ -3932,6 +3932,135 @@ FIRRTLType MultibitMuxOp::inferReturnType(ValueRange operands,
...
+
+  // verify that the class exists.
+  auto classOp = dyn_cast_or_null<ClassOp>(
+      symbolTable.lookupSymbolIn(circuitOp, className));
+  if (!classOp)
+    return emitOpError() << "target class '" << className.getValue()
+                         << "' not found";
+
+  // verify that the class type agrees with the class definition.
+  if (failed(classOp.verifyType(classType, [&]() { return emitOpError(); })))
```
Comment (dtzSiFive) [inline lib/Dialect/FIRRTL/FIRRTLOps.cpp:4047]: Nice!

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp [hunk 3]
```diff
@@ -3932,6 +3932,135 @@ FIRRTLType MultibitMuxOp::inferReturnType(ValueRange operands,
...
+    return failure();
+
+  // verify that the field exists.
```
Comment (dtzSiFive) [inline lib/Dialect/FIRRTL/FIRRTLOps.cpp:4050]:
```suggestion
  // Verify that the field exists.
```

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp [hunk 4]
```diff
@@ -3932,6 +3932,135 @@ FIRRTLType MultibitMuxOp::inferReturnType(ValueRange operands,
...
+  auto index = getIndex();
+  if (classType.getNumElements() < index)
```
Comment (dtzSiFive) [inline lib/Dialect/FIRRTL/FIRRTLOps.cpp:4052]:
```suggestion
  if (classType.getNumElements() <= index)
```

Although this should be checked already by the normal verifier, probably the check below as well can be dropped?
Don't feel strongly so if it seems useful for some reason happy to keep it after minor fixups :+1: .

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp [hunk 5]
```diff
@@ -3932,6 +3932,135 @@ FIRRTLType MultibitMuxOp::inferReturnType(ValueRange operands,
...
+    return emitOpError() << "field index does not ";
```
Comment (dtzSiFive) [inline lib/Dialect/FIRRTL/FIRRTLOps.cpp:4053]: (incomplete)

File: include/circt/Dialect/FIRRTL/FIRRTLExpressions.td [hunk 6]
```diff
@@ -1029,6 +1029,34 @@ def BitCastOp: FIRRTLOp<"bitcast", [Pure]> {
...
 
+/// The object.subfield op is different from other expressions, because its
+/// results correspond to ports, which can have foreign (not firrtl) types.
+/// For this reason, The object.subfield op is incompatible with the
+/// FIRRTLExprOp base class.
+def ObjectSubfieldOp : FIRRTLOp< "object.subfield",
+  [HasCustomSSAName, Pure,
+   DeclareOpInterfaceMethods<InferTypeOpInterface>,
+   DeclareOpInterfaceMethods<SymbolUserOpInterface>]> {
+  let summary = "Extract an element of an instance of a module";
```
Comment (dtzSiFive) [inline include/circt/Dialect/FIRRTL/FIRRTLExpressions.td:1040]:
```suggestion
  let summary = "Extract an element of an object";
```
(instance of a class)

**Review by dtzSiFive — COMMENTED** [review]
Looking good! Added some comments.  A test showing it in IR and some error checks would be great :+1: .

File: include/circt/Dialect/FIRRTL/FIRRTLExpressions.td [hunk 7]
```diff
@@ -1029,6 +1029,34 @@ def BitCastOp: FIRRTLOp<"bitcast", [Pure]> {
...
+    ```
+  }];
+  let arguments = (ins ClassType:$input, I32Attr:$index);
+  let results = (outs AnyType:$result);
+  let hasCustomAssemblyFormat = true;
+  let extraClassDeclaration = [{
+    /// Infer the return type of this operation.
+    /// Note: In contrast to other ops, this function infers a generic Type,
+    /// in order to support foreign types in ports.
+    static Type inferReturnType(ValueRange operands,
```
[the remark this answers is not in the document]
Comment (rwy7) [reply include/circt/Dialect/FIRRTL/FIRRTLExpressions.td:1055]: Yes, it is for `emitCachedSubAccess` in the parser. I've made the general `inferReturnTypes` call into the simpler `inferReturnType`.

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp (hunk 4 above)
Comment (rwy7) [reply lib/Dialect/FIRRTL/FIRRTLOps.cpp]: I was doing all the verification in verify symbol uses, I've moved this check to the verifier, and also "the result type of this op agrees with the type of the field".

File: include/circt/Dialect/FIRRTL/FIRRTLExpressions.td [hunk 8]
```diff
@@ -1029,6 +1029,34 @@ def BitCastOp: FIRRTLOp<"bitcast", [Pure]> {
 // Properties
 //===----------------------------------------------------------------------===//
 
+/// The object.subfield op is different from other expressions, because its
+/// results correspond to ports, which can have foreign (not firrtl) types.
+/// For this reason, The object.subfield op is incompatible with the
+/// FIRRTLExprOp base class.
+def ObjectSubfieldOp : FIRRTLOp< "object.subfield",
+  [HasCustomSSAName, Pure,
```
[the remark this answers is not in the document]
Comment (rwy7) [reply include/circt/Dialect/FIRRTL/FIRRTLExpressions.td]: Done, thanks

**Review by mikeurbach — APPROVED** [review]
I have a nitpicky comment about the op naming / description. It is not crucial, and I think the actual implementation looks great.

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp [hunk 9]
```diff
@@ -3932,6 +3930,120 @@ FIRRTLType MultibitMuxOp::inferReturnType(ValueRange operands,
 
   return type_cast<FIRRTLType>(operands[1].getType());
 }
+//===----------------------------------------------------------------------===//
```
Comment (youngar) [inline lib/Dialect/FIRRTL/FIRRTLOps.cpp:3933]: add new line above

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp [hunk 10]
```diff
@@ -3932,6 +3930,120 @@ FIRRTLType MultibitMuxOp::inferReturnType(ValueRange operands,
...
+// ObjectSubfieldOp
+//===----------------------------------------------------------------------===//
+
+LogicalResult ObjectSubfieldOp::inferReturnTypes(
+    mlir::MLIRContext *context, std::optional<mlir::Location> location,
+    ::mlir::ValueRange operands, ::mlir::DictionaryAttr attributes,
+    ::mlir::OpaqueProperties properties, ::mlir::RegionRange regions,
+    ::llvm::SmallVectorImpl<::mlir::Type> &inferredReturnTypes) {
```
Comment (youngar) [inline lib/Dialect/FIRRTL/FIRRTLOps.cpp:3941]: Don't need to qualify these types

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp [hunk 11]
```diff
@@ -3932,6 +3930,120 @@ FIRRTLType MultibitMuxOp::inferReturnType(ValueRange operands,
...
+  auto attrs = attributes ? attributes.getValue() : ArrayRef<NamedAttribute>{};
```
Comment (youngar) [inline lib/Dialect/FIRRTL/FIRRTLOps.cpp:3942]: Can `attributes` actually be null? (guessing no)

File: include/circt/Dialect/FIRRTL/FIRRTLExpressions.td [hunk 12]
```diff
@@ -1029,6 +1029,35 @@ def BitCastOp: FIRRTLOp<"bitcast", [Pure]> {
...
+// For this reason, the object.subfield op is incompatible with the
+// FIRRTLExprOp base class.
+def ObjectSubfieldOp : FIRRTLOp<"object.subfield",
+  [Pure,
+   DeclareOpInterfaceMethods<InferTypeOpInterface>,
+   DeclareOpInterfaceMethods<SymbolUserOpInterface>]> {
+  let summary = "Extract an element of an object";
+  let description = [{
+    The object.subfield expression refers to a subelement of an object.
+    ```
```
Comment (youngar) [inline include/circt/Dialect/FIRRTL/FIRRTLExpressions.td:1043]:
```suggestion
    ```mlir
```

**Review by youngar — APPROVED** [review]
Nits, would be great to get some error tests for ObjectSubfieldOp verifiers.

**Comment by mikeurbach** [comment]
I didn't mention it before, but I also echo the comments about adding success/failure tests for the new op.

File: include/circt/Dialect/FIRRTL/FIRRTLExpressions.td [hunk 13]
```diff
@@ -1029,6 +1029,34 @@ def BitCastOp: FIRRTLOp<"bitcast", [Pure]> {
...
+/// results correspond to ports, which can have foreign (not firrtl) types.
+/// For this reason, The object.subfield op is incompatible with the
+/// FIRRTLExprOp base class.
+def ObjectSubfieldOp : FIRRTLOp< "object.subfield",
+  [HasCustomSSAName, Pure,
+   DeclareOpInterfaceMethods<InferTypeOpInterface>,
+   DeclareOpInterfaceMethods<SymbolUserOpInterface>]> {
+  let summary = "Extract an element of an instance of a module";
+  let description = [{
+    The object.subfield expression refers to a subelement of an object.
```
[the remark this answers is not in the document]
Comment (mikeurbach) [reply include/circt/Dialect/FIRRTL/FIRRTLExpressions.td:1042]: Ok, we just discussed this synchronously, and the current naming makes sense to me now. The type of a firrtl.instance is BundleType, BundleType has elements, and individual elements are accessed from a firrtl.instance with subfield ops. Analogously, the type of a firrtl.object is ClassType, ClassType has elements, and individual elements are accessed from a firrtl.object with subfield ops. All good from my end.

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp [hunk 14]
```diff
@@ -3928,6 +3926,119 @@ FIRRTLType MultibitMuxOp::inferReturnType(ValueRange operands,
...
+ObjectSubfieldOp::verifySymbolUses(SymbolTableCollection &symbolTable) {
+  auto circuitOp = getOperation()->getParentOfType<CircuitOp>();
+  auto classType = getInput().getType();
+  auto className = classType.getNameAttr();
+
+  // Verify that the class exists.
+  auto classOp = dyn_cast_or_null<ClassOp>(
+      symbolTable.lookupSymbolIn(circuitOp, className));
+  if (!classOp)
+    return emitOpError() << "references unknown class " << className;
```
Comment (mikeurbach) [inline lib/Dialect/FIRRTL/FIRRTLOps.cpp:4036]: maybe add a failure test for this case?

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp [hunk 15]
```diff
@@ -3928,6 +3926,119 @@ FIRRTLType MultibitMuxOp::inferReturnType(ValueRange operands,
...
+
+  // Verify that the class type agrees with the class definition.
+  return classOp.verifyType(classType, [&]() { return emitOpError(); });
```
Comment (mikeurbach) [inline lib/Dialect/FIRRTL/FIRRTLOps.cpp:4039]: should the emitOpError include any extra diagnostic information? i'm not sure how the return value of the second argument to `verifyType` is used, but `return emitOpError()` seems like it should either include some error message, or just be `return failure()`. also- maybe add a failure test for this case?

**Review by mikeurbach — APPROVED** [review]
This looks good to me. Might be worth adding a couple extra tests to cover some of the verifier failures.

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp (hunk 14 above)
Comment (rwy7) [reply lib/Dialect/FIRRTL/FIRRTLOps.cpp:4036]: To be honest, I'm not sure how to write a failure-test for this case. The class' name is pulled from the type of the input value. Whatever op generated that value will fail to verify first.

Thinking about this more, maybe it isn't the responsibility of this op to verify the type of its input. The op that defines the input will verify that its result is valid. We don't need to recheck it here.

Any dissenting thoughts? I think I can just remove this code.

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp (hunk 15 above)
Comment (rwy7) [reply lib/Dialect/FIRRTL/FIRRTLOps.cpp:4039]: The second argument is a callback used to get an InFlightDiagnostic, and then verifyType will stream (`_<<_`) additional information to the diagnostic.

Thinking about this, the error reporting when the error is on a port is pretty bad, I think I will open a PR to clean it up. Here's a failure case that shows what that output looks like:

```mlir
firrtl.circuit "ClassTypeWrongPortDir" {
  firrtl.class @MyClass(out %str: !firrtl.string) {}
    // expected-error @below {{'firrtl.module' op port "str" has wrong direction, got in, expected out}}
  firrtl.module @ClassTypeWrongPortDir(out %port: !firrtl.class<@MyClass(in str: !firrtl.string)>) {}
}
```

File: lib/Dialect/FIRRTL/FIRRTLOps.cpp (hunk 14 above)
Comment (rwy7) [reply lib/Dialect/FIRRTL/FIRRTLOps.cpp:4036]: removed

**Pull request merged**